ddtrace/_trace/processor/__init__.py

```diff
@@ -304,7 +304,22 @@ def on_span_finish(self, span):
         # type: (Span) -> None
         with self._lock:
             self._span_metrics["spans_finished"][span._span_api] += 1
+
+            if span.trace_id not in self._traces:
```

Comment:
I kept going back and forth myself... should we do `self._traces[span.trace_id].spans.append(span)`, which then would be a trace of one which is then finished, and would cause it to flush immediately.... but.... I can't tell if that is correct or not.

we'd maybe ensure that the span does get encoded and flushed, but should we be caring about spans which we didn't see a span start for?

Reply:
I think I don't like that idea because my take is that `on_span_finish()` should be as read-only as possible but, at this point, we're already broken anyway. If the trace was supposed to have more than one span, than we're sending broken data instead of not sending any data. I suppose that's technically better?

Reply:
well, if we end up sending a single span, but the trace produces more later, it'll be fine since it all gets stitched together in the backend, we'll just send multiple payloads.

but is there a case where we can end up with a mismatch in spans being tracked and `num_finished` ? e.g. a situation where `num_finished` can never equal `len(trace.spans)` ?

that is sort of the inverse of the bug that is happening.

Reply:
Assuming we don't have a race condition somewhere, if the trace isn't in `self._trace`, and we add it, then:
* `trace.num_finished += 1` should mean `trace.num_finished` is now 1
* `trace.num_finished == len(trace.spans)` because we're creating the trace and the span with a single span in it
* `len(finished) == 1`, so `trace.num_finished == len(finished)`
* `len(trace.spans) == 0` so we delete the trace

So by the end of `on_span_finish()` , we still don't have a trace with the span's trace ID, right? So it's like that trace never existed?

Reply:
> So by the end of on_span_finish() , we still don't have a trace with the span's trace ID, right? So it's like that trace never existed?

but that is fine, if we create another span for that trace id, then we'll create it in `on_span_start`